Add struct and input for RF cavity element

cemitch99 opened the pull request:
Began implementation of RF cavity:  Created RF cavity struct, implemented read of required element inputs, and drafted example benchmark problem.

Steps to do:

- [x] implement integrator for the RF cavity reference trajectory values t and pt
- [x] implement Fourier coefficients of the TESLA RF cavity as default
- [x] implement integration required for the linear matrix elements
- [x] implement RF cavity benchmark example
- [x] add Python bindings
- [x] add user-facing docs for new element in manual

Follow-up PR:

- [ ] implement input of user-provided field data

cemitch99 committed:
Add struct for RF cavity

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Added a 4th-order (Ruth-Forest) symplectic integrator.

Used for pushing the reference particle in an RF cavity.

cemitch99 committed:
Merge branch 'add_rf_cavity' of https://github.com/cemitch99/impactx into add_rf_cavity

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 commented:
Added a 4th-order symplectic integrator, called for the special case of the RF cavity reference particle.

ax3l commented on examples/rfcavity/input_rfcavity.in:
```diff
@@ -0,0 +1,45 @@
+# Diagnostics
+###############################################################################
+diag.slice_step_diagnostics = true
+                                                                                                                                          50,0-1        Bot
```
```suggestion

```

cemitch99 committed:
Added the number of map/ref. traj. integration steps as user input.

cemitch99 committed:
Merge branch 'add_rf_cavity' of https://github.com/cemitch99/impactx into add_rf_cavity

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Added TESLA RF cavity model.

cemitch99 commented on src/particles/integrators/Integrators.cpp:
```diff
@@ -0,0 +1,163 @@
+        // specify Fourier coefficients
+        amrex::ParticleReal cos_coef [ncoef] = {};
+        amrex::ParticleReal sin_coef [ncoef] = {};
+
```
Below is a model for the 1.3 GHz 9-cell TESLA RF cavity:

https://journals.aps.org/prab/pdf/10.1103/PhysRevSTAB.3.092001

ax3l replied:
@cemitch99 should we add that reference in the example as a comment or so?

ax3l replied:
done :)

cemitch99 committed:
Modified integrator symp4 to access maps defined within an element namespace.

This allows map1 and map2 to access basis element parameters, and
keeps element-specific data localized within the element namespace.

RF cavity parameters are now passed correctly from the input file.

cemitch99 committed:
Added second-order symplectic integrator.

cemitch99 committed:
Implementation of integration of the linear map equations for
RF cavity.

cemitch99 committed:
Modified RF element to shift local z coordinate for integration.

This change is needed for tracking to work correctly through
multiple RF elements in succession.

cemitch99 commented on src/particles/Push.cpp:
```diff
@@ -138,14 +138,16 @@ namespace detail
+
+                        // push reference particle in global coordinates
+                        element(ref_part);
+
```
Note:  For each step, it is now essential to push the reference particle _before_ pushing the beam.  This is because "element(ref_part)" integrates the reference trajectory and the linear map equations simultaneously, the latter which are used to push the beam (for the case of the RF cavity).  This change of order should not affect any elements other than the RF cavity.

cemitch99 committed:
Add updated RF cavity example, change to ImpactX.cpp.

ax3l committed:
Merge remote-tracking branch 'mainline/development' into add_rf_cavity

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l committed:
Initial Cleaning

ax3l committed:
CMake: RF-Cavity Example

ax3l committed:
Python: RFCavity Example

ax3l committed:
RFCavity: Add `.ds()` Memberfunc

ax3l committed:
Add Missing Include

ax3l committed:
Reorder arguments

alignment still not ideal, should put map first

ax3l committed:
zeval: self-assignment `-Wself-assign`

ax3l commented on src/particles/elements/RFCavity.H:
```diff
@@ -0,0 +1,425 @@
+        amrex::ParticleReal m_Escale; //! scaling factor for RF electric field
+        amrex::ParticleReal m_freq; //! RF frequency in Hz
+        amrex::ParticleReal m_phase; //! RF driven phase in deg
+        amrex::ParticleReal m_Bzsol; //! strength of solenoid field in T
```
Is this missing in the implementation? Currently, all external solenoid fields are ignored:
```
private field 'm_Bzsol' is not used [-Werror,-Wunused-private-field]
```

ax3l commented on src/particles/elements/RFCavity.H:
```diff
@@ -0,0 +1,425 @@
+
+               // push the reference particle
+
+               auto[ez,ezp,ezint] = RF_Efield(zeval);
```
We can use https://en.cppreference.com/w/cpp/utility/tuple/ignore here:
```suggestion
               auto [std::ignore, ezp, std::ignore] = RF_Efield(zeval);
```

ax3l commented on src/particles/elements/RFCavity.H:
```diff
@@ -0,0 +1,425 @@
+
+               // push the reference particle
+
+               auto[ez,ezp,ezint] = RF_Efield(zeval);
```
see below

ax3l commented on src/particles/elements/RFCavity.H:
```diff
@@ -0,0 +1,425 @@
+
+               auto[ez,ezp,ezint] = RF_Efield(z);
+               zeval = z + tau;
+               auto[ezf,ezpf,ezintf] = RF_Efield(zeval);
```
We can use https://en.cppreference.com/w/cpp/utility/tuple/ignore here:
```suggestion
               auto [std::ignore, ezpf, ezintf] = RF_Efield(zeval);
```

ax3l commented on src/particles/elements/RFCavity.H:
```diff
@@ -0,0 +1,425 @@
+            amrex::Array2D<amrex::ParticleReal,1,6,1,6> const R = refpart.map;
+
+            // write the linear map
+            /*
+            for (int i=1; i<7; i++) {
+               for (int j=1; j<7; j++) {
+                  amrex::AllPrintToFile("mapout") << "R" << " " << i << " "
+                     << j << " " << R(i,j) << "\n";
+               }
+            }
+            */
```
leftover from debugging, should probably be removed

ax3l commented on src/particles/elements/RFCavity.H:
```diff
@@ -0,0 +1,425 @@
+                 refpart.t = t;
+                 refpart.pt = pt;
+               }
+               //zeval = zeval;
```
I would remove the line fully, since it will throw a warning to self-assign a value.
```suggestion
```

ax3l committed:
More Cleaning

ax3l committed:
RFCavity: Python Bindings

ax3l committed:
Remove Bzsol

Originally included with the intention of allowing a solenoid
magnetic field. However, in the end decided not to include it,
as it is not part of a pure RF cavity element.

ax3l commented on src/particles/elements/RFCavity.H:
```diff
@@ -0,0 +1,391 @@
+            // get the linear map
+            amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6> const R = refpart.map;
+
+            // push particles using the linear map
```
Maybe we can link a reference here for better understandability

ax3l commented on src/particles/elements/RFCavity.H:
```diff
@@ -0,0 +1,391 @@
+            amrex::ParticleReal cos_coef [ncoef] = {};
+            amrex::ParticleReal sin_coef [ncoef] = {};
+
+            // recall array indexing begins with 0
```
I do not understand where these numerical constants come from.

cemitch99 replied:
Will add a comment here specifying the origin of these constants.

These are Fourier coefficients specifying the on-axis longitudinal electric field in the default 9-cell TESLA RF cavity.  Although they will remain as default values, they can be superseded by user-provided input in a follow-up PR.

cemitch99 replied:
Will add a comment + reference here.

ax3l commented on src/particles/elements/RFCavity.H:
```diff
@@ -0,0 +1,391 @@
+               for(int j=1; j < ncoef; ++j)
+               {
+                 efield = efield + cos_coef[j]*cos(j*2*pi*z/zlen) +
+                      sin_coef[j]*sin(j*2*pi*z/zlen);
```
I think `sin_coef` is still uninitialized at this point

cemitch99 replied:
Doesn't line 199:

amrex::ParticleReal sin_coef [ncoef] = {};

initialize all elements of this array to 0.0_prt?

ax3l replied:
Ah yes, you are right! :)
- https://en.cppreference.com/w/cpp/language/aggregate_initialization
- https://stackoverflow.com/a/201116/2719194

I'll make this a `constexpr std::array` as a small update https://en.cppreference.com/w/cpp/container/array

ax3l committed:
Remove local string in device code

cemitch99 commented on examples/rfcavity/analysis_rfcavity.py:
```diff
@@ -0,0 +1,113 @@
+)
+
+atol = 0.0  # ignored
+rtol = num_particles**-0.5  # from random sampling of a smooth distribution
```
This level of tolerance is so tight that it is easily violated if one changes the random seed, or changes the initialization procedure by moving from 1 MPI rank to > 1 MPI rank.  This is due to the statistical distribution of the moments of the initialized beam.  I increased the tolerance to a more robust value that works when run on 1-4 MPI ranks.

ax3l replied:
Awesome, thanks a lot!

cemitch99 committed:
Update analysis_rfcavity.py

Increased test tolerance slightly.

cemitch99 commented on examples/rfcavity/analysis_rfcavity.py:
```diff
@@ -0,0 +1,113 @@
+)
+
+atol = 0.0  # ignored
+rtol = num_particles**-0.5  # from random sampling of a smooth distribution
```
See the above comment regarding tolerance values.

cemitch99 committed:
Update RFCavity.H

Added suggested comments.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l commented on src/particles/elements/RFCavity.H:
```diff
@@ -0,0 +1,403 @@
+            // Fourier coefficients for the on-axis longitudinal
+            // electric field Ez of the 9-cell TESLA superconducting
+            // cavity (default cavity model):
+            https://journals.aps.org/prab/pdf/10.1103/PhysRevSTAB.3.092001.
```
Fixing comment:
```suggestion
            // https://journals.aps.org/prab/pdf/10.1103/PhysRevSTAB.3.092001.
```

ax3l committed:
Fix comment

ax3l commented on src/particles/integrators/Integrators.H:
```diff
@@ -0,0 +1,119 @@
+        }
+    }
+
+    template <typename T_Element>
```
We need to add the "hostdevice" and "inline" macros here: 
```suggestion
    template <typename T_Element>
    AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
```

ax3l commented on src/particles/integrators/Integrators.H:
```diff
@@ -0,0 +1,119 @@
+namespace impactx::integrators
+{
+
+    template <typename T_Element>
```
```suggestion
    template <typename T_Element>
    AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
```

ax3l commented on src/particles/integrators/Integrators.H:
```diff
@@ -0,0 +1,119 @@
+        }
+    }
+
+    template <typename T_Element>
```
```suggestion
    template <typename T_Element>
    AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
```

ax3l committed:
Integrators: GPU Support

ax3l commented on src/particles/elements/RFCavity.H:
```diff
@@ -0,0 +1,391 @@
+            amrex::ParticleReal const E0 = m_escale;
+
+            // push the reference particle
+            [[maybe_unused]] auto [ez, ezp, ezint] = RF_Efield(zeval);
```
```suggestion
            amrex::ParticleReal ezp;
            std::tie(std::ignore, ezp, std::ignore) = RF_Efield(zeval);
```

ax3l commented on src/particles/elements/RFCavity.H:
```diff
@@ -0,0 +1,391 @@
+            amrex::ParticleReal const E0 = m_escale;
+
+            // push the reference particle
+            [[maybe_unused]] auto [ez, ezp, ezint] = RF_Efield(z);
```
```suggestion
            amrex::ParticleReal ezp, ezint;
            std::tie(std::ignore, ezp, ezint) = RF_Efield(z);
```

ax3l commented on src/particles/elements/RFCavity.H:
```diff
@@ -0,0 +1,391 @@
+            // push the reference particle
+            [[maybe_unused]] auto [ez, ezp, ezint] = RF_Efield(z);
+            zeval = z + tau;
+            [[maybe_unused]] auto [ezf, ezpf, ezintf] = RF_Efield(zeval);
```
```suggestion
            amrex::ParticleReal ezpf, ezintf;
            std::tie(std::ignore, ezpf, ezintf) = RF_Efield(zeval);
```

ax3l committed:
Fix unused variable warnings

ax3l commented on src/particles/integrators/Integrators.H:
```diff
@@ -0,0 +1,122 @@
+
+#include <AMReX_Extension.H>  // for AMREX_RESTRICT
+#include <AMReX_REAL.H>       // for ParticleReal
+#include <AMReX_Print.H>      // for PrintToFile
```
```suggestion
```

ax3l commented on src/particles/elements/RFCavity.H:
```diff
@@ -0,0 +1,406 @@
+            refpart.s = s + slice_ds;
+        }
+
+        std::tuple<amrex::ParticleReal,amrex::ParticleReal,amrex::ParticleReal>
```
```suggestion
        std::tuple<amrex::ParticleReal, amrex::ParticleReal, amrex::ParticleReal>
        AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
```

ax3l commented on src/particles/elements/RFCavity.H:
```diff
@@ -0,0 +1,406 @@
+            return std::make_tuple(efield,efieldp,efieldint);
+        }
+
+        void map3 (amrex::ParticleReal const tau,
```
```suggestion
        AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
        void map3 (amrex::ParticleReal const tau,
```

ax3l commented on src/particles/elements/RFCavity.H:
```diff
@@ -0,0 +1,406 @@
+            refpart.map(5,6) = R(5,6) + tau*R(6,6)/pow(betgam,3);
+        }
+
+        void map2 (amrex::ParticleReal const tau,
```
```suggestion
        AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
        void map2 (amrex::ParticleReal const tau,
```

ax3l commented on src/particles/elements/RFCavity.H:
```diff
@@ -0,0 +1,406 @@
+            refpart.map(4,4) = -s*pow(L,2)*R(3,4) + (1.0_prt+s*L)*R(4,4);
+        }
+
+        void map1 (amrex::ParticleReal const tau,
```
```suggestion
        AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
        void map1 (amrex::ParticleReal const tau,
```

ax3l committed:
RFCavity: GPU Support

ax3l committed:
GPU Support: Rewrite ignore unused

ax3l commented on src/particles/elements/RFCavity.H:
```diff
@@ -0,0 +1,411 @@
+            refpart.s = s + slice_ds;
+        }
+
+        std::tuple<amrex::ParticleReal, amrex::ParticleReal, amrex::ParticleReal>
```
Let's add a Doxygen comment for this member function.

ax3l commented on src/particles/elements/RFCavity.H:
```diff
@@ -0,0 +1,411 @@
+            return std::make_tuple(efield,efieldp,efieldint);
+        }
+
+        AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
```
Let's add a Doxygen comment for this member function.

ax3l commented on src/particles/elements/RFCavity.H:
```diff
@@ -0,0 +1,411 @@
+            refpart.map(4,4) = -s*pow(L,2)*R(3,4) + (1.0_prt+s*L)*R(4,4);
+        }
+
+        AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
```
Let's add a Doxygen comment for this member function.

ax3l committed:
Optimizations/Moderniz. on Constants

ax3l commented on src/particles/ReferenceParticle.H:
```diff
@@ -39,6 +40,9 @@ namespace impactx
         amrex::ParticleReal charge = 0.0; ///< reference charge, in C
 
+        amrex::ParticleReal sedge = 0.0;  ///< value of s at entrance of the current beamline element
+        amrex::Array2D<amrex::ParticleReal, 1, 6, 1, 6> map; ///< linearized map
```
I am ok to have these 1-based, close to the mathematical formula.
But I think the compiler might be able to simplify a few integer operations when we access the arrays at runtime (offsets and then modulo operations) if we had it starting at zero.

Although these index accesses appear to be known at compile-time the way we access it (`R(2,3)`, etc.), so it might not matter.

cc @WeiqunZhang ?

WeiqunZhang replied:
I agree 1-based indexing should have the same speed as 0-based because it's a known constant at compile time.

ax3l committed:
RFCavity Element: User-Facing Docs

cemitch99 committed:
Add, modify documentation related to RF cavity functions.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ax3l reviewed (approved):
Thank you, this looks great! :rocket: :sparkles:
